Update URLs, corrections, suggest wording improvements (#86689)
Azure/azure-content: articles/active-directory-b2c/openid-connect.md

When your web application needs to authenticate the user and run a user flow, it can direct the user to the `/authorize` endpoint. The user takes action depending on the user flow.

> In this request, the client indicates the permissions that it needs to acquire from the user in the `scope` parameter, and specifies the user flow to run. To get a feel for how the request works, try pasting the request into a browser and running it. Replace `{tenant}` with the name of your tenant. Replace `90c0fe63-bcf2-44d5-8fb7-b8bbc0b29dc6` with the app ID of an application you've previously registered in your tenant or [register a web application](tutorial-register-applications.md#register-a-web-application). Replace `{application-id-uri}/{scope-name}` with the **Application ID URI** and **Scope** of an application that you've previously registered in your tenant. Go to the Azure AD B2C blade, select **App registrations**, select **All applications** and select an application. Under the **Manage** section, select **Expose an API**. If it's preferable to start fresh, [register a new web api](add-web-api-application.md#app-registrations) and [configure scopes](add-web-api-application.md#configure-scopes). Also change the policy name (`{policy}`) to the policy name that you have in your tenant, for example `b2c_1_sign_in`.

Reviewer: ```suggestion
In this request, the client indicates the permissions that it needs to acquire from the user in the `scope` parameter, and specifies the user flow or policy to run. To get a feel for how the request works, paste the request into your browser and run it. Replace:
-  `{tenant}` with the name of your tenant.
-  `90c0fe63-bcf2-44d5-8fb7-b8bbc0b29dc6` with the app ID of an [application you've previously registered in your tenant](tutorial-register-applications.md).  
- `{application-id-uri}/{scope-name}` with the Application ID URI and Scope of an application that you've previously registered in your tenant
- `{policy}` with the policy name that you have in your tenant, for example `b2c_1_sign_in`.
```

Author: Thanks for the review, both corrections and the list will make the section more readable.

```http
GET https://{tenant}.b2clogin.com/{tenant}.onmicrosoft.com/{policy}/oauth2/v2.0/authorize?
client_id=90c0fe63-bcf2-44d5-8fb7-b8bbc0b29dc6

> &response_type=code+token

Reviewer: ```suggestion
&response_type=code+id_token
```

Reviewer: at this point, we're interested in id_token to identify a user, we're not calling any API. Let's retain original content.

Author: Ah!
I didn't know about "hybrid flow" though the the code+id_token response type is not new to me, but I haven't used it and so I made an incorrect assumption that there was no use for it in the example either.
I will also add that the id_token that is returned when using response_type=code+id_token results in an ID token that when pasted in jwt.ms it isn't decoded.
jwt.io does try, and I can see both aud and sub claims, but it also tells me that the token is incomplete because of an invalid signature, which explains why jwt.ms "discards" it (but the subsequent ID token, post code exchange, is OK).

This is only to try to explain why I didn't understand why it was included.
I just think this might be a bit confusing for other users that read and test the examples.
But I am guessing this is something that can be tackled in another PR or another article :-)

Now I have learned about the hybrid flow and use-cases for it and I understand why it's part of the sample request.

&redirect_uri=https%3A%2F%2Fjwt.ms%2F
&response_mode=fragment
&scope=openid%20offline_access%20{application-id-uri}/{scope-name}
&state=arbitrary_data_you_can_receive_in_the_response
&nonce=12345
```

| Parameter | Required | Description |
| --------- | -------- | ----------- |
| {tenant} | Yes | Name of your [Azure AD B2C tenant](tenant-management.md#get-your-tenant-name). If you're using a [custom domain](custom-domain.md), replace `tenant.b2clogin.com` with your domain, such as `fabrikam.com`.  |
| {policy} | Yes | The user flow to be run. Specify the name of a user flow you've created in your Azure AD B2C tenant. For example: `b2c_1_sign_in`, `b2c_1_sign_up`, or `b2c_1_edit_profile`. |
| client_id | Yes | The application ID that the [Azure portal](https://portal.azure.com/) assigned to your application. |
| nonce | Yes | A value included in the request (generated by the application) that is included in the resulting ID token as a claim. The application can then verify this value to mitigate token replay attacks. The value is typically a randomized unique string that can be used to identify the origin of the request. |

> | response_type | Yes | The response type, which must include `code` for the authorization code flow. If your web application also needs tokens for calling a web API, you can use `code+token`. When adding `token` to the response_type parameter value you must add a valid resource in the `scope` parameter. |

Reviewer: ```suggestion
| response_type | Yes | Must include an ID token for OpenID Connect. If your web application also needs tokens for calling a web API, you can use `code+id_token`. |
```

Reviewer: retain original content!

Author: As mentioned earlier, I understand why we better retain the original content.
We are not changing anything that does not need to be changed!

| scope | Yes | A space-separated list of scopes. The `openid` scope indicates a permission to sign in the user and get data about the user in the form of ID tokens. The `offline_access` scope is optional for web applications. It indicates that your application will need a *refresh token* for extended access to resources. The `{application-id-uri}/{scope}` indicates a permission to protected resources, such as a web API. For more information, see [Request an access token](access-tokens.md#scopes). |
| redirect_uri | Yes | The `redirect_uri` parameter of your application, where authentication responses can be sent and received by your application. It must exactly match one of the `redirect_uri` parameters that you registered in the Azure portal, except that it must be URL-encoded. |
| response_mode | Recommended | The method that is used to send the resulting authorization code back to your application. It can be either `query`, `form_post`, or `fragment`.  The `form_post` response mode is recommended for best security. |

> | state | Recommended | A value included in the request that's also returned in the token response. It can be a string of any content that you want. A randomly generated unique value is typically used for preventing cross-site request forgery attacks. The state is also used to encode information about the user's state in the application before the authentication request occurred, such as the page they were on. |

Reviewer: ```suggestion
| state | Recommended | A value included in the request that's also returned in the token response. It can be a string of any content that you want. A randomly generated unique value is typically used for preventing cross-site request forgery attacks. The state is also used to encode information about the user's state in the application before the authentication request occurred, such as the page they were on. If you don't want to register multiple redirect URLs in your Azure portal, you can use the `state` parameter to differentiate responses in your application from Azure AD B2C service due to different requests. |
```

Author: Nice!
For developers new to authentication or with limited experience, more ideas on how it's possible to be creative within an authentication exchange is really good.

After the user completes the user flow, a response is returned to your application at the indicated `redirect_uri` parameter, by using the method that's specified in the `response_mode` parameter. The response is the same for each of the preceding cases, independent of the user flow.

> A successful response using `response_mode=fragment` and `response_type=code+token` would look like:

Reviewer: ```suggestion
A successful response using `response_mode=fragment` would look like:
```

Reviewer: Let's retain original content!

Author: Agreed 💯

```http
GET https://jwt.ms/#
access_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6Ik5HVEZ2ZEstZnl0aEV1Q...
&code=eyJraWQiOiJBQjk4dkdFRHY2Nl9ScmJEajJmclFFMUFOM1hSX3k4ZVU3X3V5NDBjN...

> &state=arbitrary_data_you_can_receive_in_the_response

Reviewer: On the response, only update the GET URL to https://jwt.ms/#, leave everything else as is. We won't receive an access token from `/authorize` endpoint using code+id_token response type.

```

| Parameter | Description |
| --------- | ----------- |

> | access_token | The access token for a target resource that the application requested if you used `response_type=code+token`. |

Reviewer: ```suggestion

```

Reviewer: No access_token in the response

Author: Yes, with response_type=code+id_token, the access token is not requested and should not be mentioned 👍

## Get a token

If you need your web application to only run user flows, you can skip the next few sections. These sections are applicable only to web applications that need to make authenticated calls to a web API which is also protected by Azure AD B2C.

> You can redeem the authorization code that you acquired (by using `response_type=code`) for a token to the desired resource by sending a `POST` request to the `/token` endpoint. In Azure AD B2C, you can [request access tokens for other APIs](access-tokens.md#request-a-token) as usual by specifying their scope(s) in the request.

Reviewer: ```suggestion
You can redeem the authorization code that you acquired (by using `response_type=code+id_token`) for a token to the desired resource by sending a `POST` request to the `/token` endpoint. In Azure AD B2C, you can [request access tokens for other APIs](access-tokens.md#request-a-token) as usual by specifying their scope(s) in the request.
```

Author: Yes, for consistency, we must include +id_token.

> You can also request an access token for your app's own back-end Web API by convention of using the app's **Application ID URI** plus **Scope** name (`{application-id-uri}/{scope-name}`) as the requested scope (which will result in an access token with the desired resource's client ID as the "audience"):

Reviewer: ```suggestion
You can also request an access token for your app's own back-end Web API by convention of using the app's client ID as the requested scope (which will result in an access token with that client ID as the "audience"):
```

Reviewer: Do not make any change here! You use app ID (no need of scope) when an app wants to access an api of itself.

Author: Agreed!
I got to caught up in the first leg of the authorization_code flow to even think that this section mentions when the application is calling itself as a backend API (and regardless if that is best practice or not, it's certainly true that it works) 👍 
Thanks for saving me 🥇  from publishing incorrect content, when I tested I was another app registration as API...
